general/mem_manager.hpp

```diff
@@ -67,119 +496,268 @@ class MemoryManager
-   void *GetDevicePtr(const void *ptr);
+// Inline methods
 
-   /// Registers external host pointer in the memory manager which will manage
-   /// the corresponding device pointer, but not the provided host pointer.
-   template<class T>
-   void RegisterHostPtr(T *ptr_host, const std::size_t size)
+template <typename T>
+inline void Memory<T>::New(int size, MemoryType mt)
+{
+   if (mt == MemoryType::HOST)
    {
-      Insert(ptr_host, size*sizeof(T));
-#ifdef MFEM_DEBUG
-      RegisterCheck(ptr_host);
-#endif
+      New(size);
    }
+   else
+   {
+      // Allocate the host pointer with new T[] if 'mt' is a pure device memory
+      // type, e.g. CUDA.
+      T *tmp = (mt == MemoryType::CUDA) ? new T[size] : NULL;
+      h_ptr = (T*)MemoryManager::New_(tmp, size*sizeof(T), mt, flags);
+      capacity = size;
+   }
+}
 
-   /// Registers external host and device pointers in the memory manager.
-   template<class T>
-   void RegisterHostAndDevicePtr(T *ptr_host, T *ptr_device,
-                                 const std::size_t size, const bool host)
+template <typename T>
+inline void Memory<T>::Wrap(T *ptr, int size, MemoryType mt, bool own)
+{
+   if (mt == MemoryType::HOST)
+   {
+      Wrap(ptr, size, own);
+   }
+   else
    {
-      RegisterHostPtr(ptr_host, size);
-      SetHostDevicePtr(ptr_host, ptr_device, host);
+      // Allocate the host pointer with new T[] if 'mt' is a pure device memory
+      // type, e.g. CUDA.
+      T *tmp = (mt == MemoryType::CUDA) ? new T[size] : NULL;
+      h_ptr = (T*)MemoryManager::Register_(ptr, tmp, size*sizeof(T), mt, own,
+                                           false, flags);
+      capacity = size;
    }
+}
 
-   /// Set the host h_ptr, device d_ptr and mode host of the memory region just
-   /// been registered with h_ptr (see RegisterHostAndDevicePtr)
-   void SetHostDevicePtr(void *h_ptr, void *d_ptr, const bool host);
+template <typename T>
+inline void Memory<T>::MakeAlias(const Memory &base, int offset, int size)
+{
+   h_ptr = base.h_ptr + offset;
+   capacity = size;
+   if (!(base.flags & REGISTERED))
+   {
+      flags = (base.flags | ALIAS) & ~(OWNS_HOST | OWNS_DEVICE);
+   }
+   else
+   {
+      MemoryManager::Alias_(base.h_ptr, offset*sizeof(T), size*sizeof(T),
+                            base.flags, flags);
+   }
+}
 
-   /// Unregisters the host pointer from the memory manager. To be used with
-   /// memory not allocated by the memory manager.
-   template<class T>
-   void UnregisterHostPtr(T *ptr) { Erase(ptr); }
+template <typename T>
+inline void Memory<T>::Delete()
+{
+   if (!(flags & REGISTERED) ||
+       MemoryManager::Delete_(h_ptr, flags) == MemoryType::HOST)
+   {
+      if (flags & OWNS_HOST) { delete [] h_ptr; }
+   }
+}
 
-   /// Check if pointer has been registered in the memory manager
-   void RegisterCheck(void *ptr);
+template <typename T>
+inline T &Memory<T>::operator[](int idx)
+{
+   MFEM_ASSERT((flags & VALID_HOST) && !(flags & VALID_DEVICE),
```

Comment:
There seem to be different assertions across the `Memory` class, is that intended?
```cpp
1. (flags & VALID_HOST) && !(flags & VALID_DEVICE)
2. (flags & VALID_HOST)
3. (flags & VALID_HOST) && (std::is_const<T>::value || !(flags & VALID_DEVICE))
```

Reply:
Yes! For example, if the method allows only read access (i.e. it returns a reference to a const type) the we only care if the host pointer is valid (reading the value will not change the validity of the host or device pointers). However, if the method allows write access (i.e. the return type is a reference to non-const type) then we want to make sure the device pointer is not marked as valid because overwriting the value in host memory will render the `VALID_DEVICE` flag incorrect.

These asserts are active only when `MFEM_DEBUG=YES` -- they are a debugging tool that will tell you if you are trying to use the host pointer incorrectly.